### ISSUE TITLE
About DAG networks

### ISSUE THREAD

**Issue by tinalegre** [vlfeat/matconvnet#314]
I was wondering if any of you has been using the **DAG** module of matconvnet? I have been mainly training **linear CNNs** and would like to train now **DAG networks**. Could anyone share their experiences and the network architectures you have been defining? I found the matconvnet examples on training/testing linear CNNs very useful (such as **cnn_image_net**, **cnn_cifar**, etc.), but found some lack of examples for DAG architectures.
It would be nice to hear how matconvnet has been used for such sort of architectures.
I'm especially interested in: (i) parallel linear CNNs which fuse together into a single FC layer followed
by other(s) FC layers, or (ii) linear CNNs whose before last layer branches out into different subnets.
Any experiences/advices are more than welcome!

Below I'm illustrating two DAG networks. The first illustration consists of 3 fine-tunned linear CNNs (in blue, orange, and green) which fuse  together into a single FC layer (red):
![First multi-parallel architecture](http://i.imgur.com/zBhvIQI.png?2)

The second example is adapted from the PANDA paper (Zhang et. al. CVPR 2014), in which a FC layer (e.g. FC_7) branches into different FC layers (shown in red).
![First multi-output architecture](http://i.imgur.com/sZTIG4S.png?1)

**Comment by korkinof** [comment]
Well, I raised the same issue a few weeks ago and the response was that "no more examples are necessary" and the issue was closed.
Now, as I was interested in using matconvnet at the time, I looked into it in detail.
The ONLY example available is converting a simplenn to a dagnn, which however does not involve concatenation or batchnormalisation. If there were an example, using those two important elements, it would have immediately become apparent that those layers do not work properly.
Therefore, apparently, the single example provided is insufficient. Hopefully the present issue will not be closed until it is actually resolved. Otherwise, we can just keep raising it.

**Comment by bazilas** [comment]
Hi @tinalegre,

I have switched to the DagNN structure a few weeks ago and had no problem to implement dag architectures or transfer my current ones to dag networks.

It could take some time to understand the DagNN structure, but in general it is straightforward to write configurations with multiple losses and concatenations. Maybe have a look on the googlenet example. The function fromSimpleNN() might be helpful as well. 

cheers,
Vasilis

**Comment by nicjac** [comment]
@korkinof do you have any example of the issues you encountered with concatenation and batch normalisation? I have been using both and I do not see anything that would suggest they do not work properly / as intended.

**Comment by korkinof** [comment]
@nicjac Sure, take a look at this issue:
https://github.com/vlfeat/matconvnet/issues/280

**Comment by vedaldi** [comment]
Just to be clear:

- beta16 applies batch normalisation at training and test time. This is not *wrong*, but it is probably a behaviour which is *inconvenient* at test time, for two reasons:

1) batch normalisation introduces noise, which is good for training (see dropout) but bad for testing (as it yields lower accuracies)
2) batch normalisation requires a stack of images to operate on, so it would not work for testing on a single image

- with a little effort, is absolutely possible to address 1) and 2) in beta16. After training, simply compute the average moments on a few images and multiply-add those in to the filters after batch normalisation, then remove the layer. Admittedly, however, it is inconvenient.

- beta17, currently in the works, will dramatically simplify all this as it stores the moments during training, therefore allowing to bypass batch normalisation at test time with a switch

Concerning concatenation, there was a bug in backpropagation in the first beta release that contained that layer. It has been fixed many weeks ago and I am not aware of further bugs.

Best wishes,

Andrea

**Comment by anewlearner** [comment]
@tinalegre  Have you solved the problem. I looked through the discussion,but still don't know how to implement the first dagnn model above.
Examples would be appreciated. Thanks!

**Comment by anewlearner** [comment]
@bazilas  Just as what you suggested, I have looked through the googlenet example. But the first dagnn example in the issue is different from the googlenet which only has one input layer but concatenate different layers in the same net. So in googlenet, we can specify the input layer and output layer in 'net.addLayer()' to realize concatenation.
But the case here is that we have several input layers(for net) such as multi-scale or multi-view images which means that there are parallel nets and need to concatenate layers in the different nets,as shown in the figure above.
I'm not sure whether my understanding is correct or not. Hope you can help. Thanks!

**Comment by bazilas** [comment]
In dagNN, you can have several inputs. It works fine. Having also several networks can be modelled in dagNN. I will try to upload an example as soon as possible.

**Comment by anewlearner** [comment]
Thanks for your quick answer. Look forward for your example. :)

**Comment by IamKitty** [comment]
@bazilas   I have troulbe in training a composite DNN which includes two sub-CNNs.  Would you be so kind as to upload your DagNN example? Thanks a lot.

**Comment by yushuinanrong** [comment]
@vedaldi @bazilas 
Indeed, some examples showing how to use DAG network for models like (i) and (ii) are extremely helpful. The filed of CNN are pushing more and more complicate architectures, DAG networks would be used more I guess.

Thanks

**Comment by alex-lijun** [comment]
hi all, 
when i am trying to run the " cnn_imagenet_googlenet", i always meet this problem:Error using dagnn.Layer/load (line 191)
No property `dilate` for a layer of type `dagnn.Conv`.

Error in dagnn.DagNN.loadobj (line 28)
      block.load(struct(s.layers(l).block)) ;

Error in cnn_imagenet_googlenet (line 14)
net = dagnn.DagNN.loadobj(load(modelPath)) ;

what should i do ?

**Comment by fengyunxiaozi** [comment]
hi, I was currently using DAG wrapper and  also wanted to implement networks as shown in above figures. Have you uploaded your examples? @bazilas

**Comment by jesszgc** [comment]
to finetue googlenet i got the same problem like yours @alex-lijun

**Comment by micklexqg** [comment]
@jesszgc, I got the same problem when finetune googlenet, have you solved it?

**Comment by Hosseinka** [comment]
I got the same issue as @alex-lijun and others, has anyone solved this error ???

**Comment by kilakila-heart** [comment]
to finetue wsdnn i got the same issue as @alex-lijun and others, has anyone solved this error ???
